pkg/fswatcher/fswatcher.go

```diff
@@ -133,12 +139,20 @@ func (w *FSWatcher) watch() {
 				return
 			}
 			w.logger.Error("fsnotifier reported an error", zap.Error(err))
+		case <-w.stop:
+			w.wg.Done()
+			return
 		}
 	}
 }
 
-// Close closes the watcher.
+// Close closes the watcher and stops the background go routine of FSWatcher.
 func (w *FSWatcher) Close() error {
+	if w.isClosed.CompareAndSwap(false, true) {
```

Comment:
I was looking closer at this and I am not convinced this is all needed. When we call `w.watcher.Close()`, it closes the `Events` and `Errors` channels, so our goroutine will exit naturally. Yes, there is a race condition if you're checking for its exit immediately, but that cannot be completely avoided, even with these changes the goroutine may go to sleep after L143 lowers the semaphore.

If you observed that the goroutine is still running for a long time after close, there may be something else going on.

We can still use the waitgroup introduced here, but without the channel/bool:
```go
func (w *FSWatcher) Close() error {
	// watcher.Close() takes care of multiple calls by returning "already closed" error
	if err := w.watcher.Close(); err != nil {
		return err
	}
	w.wg.Wait()
	return nil
}
```

Reply:
I see. let me send out a new commit with these changes.

Reply:
Can you elaborate how you determined that the GR was still running? Can it be reproduced with a unit test?

Reply:
+1